include/cuco/bloom_filter.cuh

```diff
@@ -0,0 +1,376 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/bloom_filter_policy.cuh>
+#include <cuco/bloom_filter_ref.cuh>
+#include <cuco/detail/storage/storage_base.cuh>
+#include <cuco/extent.cuh>
+#include <cuco/hash_functions.cuh>
+#include <cuco/utility/allocator.hpp>
+#include <cuco/utility/cuda_thread_scope.cuh>
+
+#include <cuda/atomic>
+#include <cuda/std/array>
+#include <cuda/std/cstddef>
+#include <cuda/stream_ref>
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+
+namespace cuco {
+
+/**
+ * @brief A GPU-accelerated Blocked Bloom filter.
+ *
+ * The `bloom_filter` supports two types of operations:
+ * - Host-side "bulk" operations
+ * - Device-side "singular" operations
+ *
+ * The host-side bulk operations include `add`, `contains`, etc. These APIs should be used when
+ * there are a large number of keys to add or lookup. For example, given a range of keys
+ * specified by device-accessible iterators, the bulk `add` function will add all keys into
+ * the filter.
+ *
+ * The singular device-side operations allow individual threads (or Cooperative Groups) to perform
+ * independent add or lookup operations from device code. These operations are accessed through
+ * non-owning, trivially copyable reference types (or "ref").
+ *
+ * @tparam Key Key type
+ * @tparam Extent Size type that is used to determine the number of blocks in the filter
+ * @tparam Scope The scope in which operations will be performed by individual threads
+ * @tparam Policy Type that defines how to generate and store key fingerprints (see
+ * `cuco/bloom_filter_policy.cuh`)
+ * @tparam Allocator Type of allocator used for device-accessible storage
+ */
+template <class Key,
+          class Extent             = cuco::extent<std::size_t>,
+          cuda::thread_scope Scope = cuda::thread_scope_device,
+          class Policy =
+            cuco::bloom_filter_policy<cuco::xxhash_64<Key>, cuda::std::array<std::uint32_t, 8>>,
```

Comment:
Hmm, I also thought about this. The information needed here is the word type and block size (determined by `typename Block::value_type` and `cuda::std::tuple_size_v<Block>`, i.e., standardized facilities to describe a container's value type and static size). So technically any container that follows this concept can be used to describe the filter block parameters. 

This type has a single use in cuco and since the `Block` type itself is actually never used, i.e., we need it only to extract the block parameters, I would rather vote for replacing the `Block` tparam in the policy with `bloom_filter_policy<class Hash, class Word, uint32_t WordsPerBlock>`. WDYT?